client/src/contexts/boardContext.js

```diff
@@ -0,0 +1,100 @@
+import React, {createContext, useEffect, useReducer, useState} from 'react';
+import Board from '../api/Board';
+ 
+
+const initialState = {
+    id: null,
+    columns: {},
+    columnOrder: [],
+    tasks: {}
+}
+const BoardContext = createContext(initialState)
+
+const reducers = {
+    addCard: (state, task, colId) => {
+        const newState = {...state}
+        newState.tasks[task.id] = {id: task.id, content: task.name}
+        newState.columns[colId].taskIds.push(task.id)
+        return newState;
+    },
+    moveCard: (state, prevCol, nextCol, oldIndex, newIndex) => {
+        const newState = {...state}
+        const taskId = newState.columns[prevCol].taskIds[oldIndex]
+        newState.columns[prevCol].taskIds.splice(oldIndex, 1)
+        newState.columns[nextCol].taskIds.splice(newIndex, 0, taskId)
+        return newState;
```

Comment:
can you try something like:
```
const taskId = state.columns[prevCol].taskIds[oldIndex];
const prevColumn = state.columns[prevCol].filter((card, idx) => return idx !== oldIndex);
const newColumn = [...state.columns[nextCol]]
newColumn.taskIds.splice(newIndex, 0, taskId)
//it's an array right?
const newColumnsArray = state.columns.map((column, idx) => {
if(idx === prevCol) return prevColumn;
if(idx === nextCol) return newColumn;
return column
}
const newState = {...state, columns: newColumnsArray}
return newState
```

So this seems a little overcomplicated (it's also messy can defs be cleaned up), but when you do {...state} it does a shallow copy and keeps a lot of the arrays within the same. Then when you directly edit it with splice you're also not creating new arrays. So once you set the state, React does a shallow check of the state to see if anything has changed and assumes that the arrays have not changed.

Reply:
Thanks for the tip! I did a variation of what you posted and the state works properly now. Guess I gotta be a little more precise on deep copying lol.